### ISSUE TITLE
[Feature Request] Swish Activation Function

### ISSUE THREAD

**Issue by isaykatsman** [pytorch/pytorch#3169]
Swish ([arxiv](https://arxiv.org/pdf/1710.05941.pdf)) is an activation function that has been shown to empirically outperform ReLU and several other popular activation functions on Inception-ResNet-v2 and MobileNet. On models with more layers Swish typically outperforms ReLU. Implementation is simple:

![swish](https://i.imgur.com/KhHvAIy.png)

Sigma is just sigmoid.

Worth a PR?

cc @albanD @mruberry

**Comment by apaszke** [comment]
Let's wait until it's actually used somewhere outside of this paper. It's so easy to implement it (`x * x.sigmoid()`) that I doubt it's worth adding so early. Thanks!

**Comment by gchanan** [comment]
From slack (@soumith):

>>>I reviewed the data, where the activation function is used now, etc. It looks appropriate to add now (edited) 
(our last review was in 2017, so understandably we didn't have enough data then)

**Comment by albanD** [comment]
I think the right way to go with this is a first step that just adds a nn.Module and nn.functional version that uses `x * x.sigmoid()`.
Then if we have performance issues, we can do a follow up PR that will implement it as a single cpp operator.

**Comment by rwightman** [comment]
@albanD I can confirm that there are performance concerns (throughput and GPU memory) with the basic Python nn.Module/nn.functional. I don't believe there is much point in adding it without going down to the cpp operator level given how simple it is to define.

I brought up the question in slack due the the fact that I and others have been working around the issue of not having an optimized Swish impl in PyTorch by doing complicated mixes of jit scripting (for the speed via fusion) and autograd fn (for memory efficiency). The improved autograd results in roughly 15-20% GPU memory savings on some of the larger EfficientNet models. The speed gains from the fusion are smaller but worthwhile on such large networks.

PyTorch 1.5/1.5.1 killed the jit fusion gains and caused a performance regression due to changes in the default executor, it'd be nice not needing to worry about those changes for common activations.

I've been working around the limitiations by building an activation factory that can switch between autograd fn, jit scripted, and plain python based on the needs (for exportable models, etc) https://github.com/rwightman/pytorch-image-models/blob/master/timm/models/layers/activations_me.py#L18

There is at least one CUDA cpp impl that some have been using https://github.com/thomasbrandon/swish-torch

**Comment by albanD** [comment]
Ho right I can see the benefit from the memory side.
I was just trying to scope this as small as possible for the person that is going to implement it.

But yes we should be able to add something very close to sigmoid on the cpp side to get an easy implementation.
Happy to help with guidance/review if needed.

**Comment by rwightman** [comment]
@albanD great, I'll try to find a bit of time to poke around and see what it'd involve, but definitely open to others picking it up too...

Between the two most common PyTorch impl of EfficientNets (and related networks), both mine and Luke's leverage a custom autograd, it's been used in countless Kaggle Challenges (all of the top solutions in the recent FB deepfake challenge) and a few real world applications by several companies (at least that I'm aware of). There've been numerous issues filed and questions about why can't I torchscript / export (ONNX) this model, etc... having an optimized version here would be very useful.

**Comment by rwightman** [comment]
@albanD I should have some time to start working on this soon, a question before I begin adding, what is the preferred ordering for inserting new changes? I don't see an obvious pattern (always insert new at the top/bottom, insert alphabetical, etc)

Random example from `Activation.h` 
```
using activation_fn = void (*)(TensorIterator&);
using activation_backward_fn = void (*)(TensorIterator&);
using softplus_fn = void (*)(TensorIterator&, Scalar, Scalar);
using softplus_backward_fn = void (*)(TensorIterator&, Scalar, Scalar);
using threshold_fn = void (*)(TensorIterator&, Scalar, Scalar);
using hardtanh_backward_fn = void (*)(TensorIterator&, Scalar, Scalar);
using hardsigmoid_fn = void(*)(TensorIterator&);
using hardsigmoid_backward_fn = void(*)(TensorIterator&);
using hardswish_fn = void(*)(TensorIterator&);
using hardswish_backward_fn = void(*)(TensorIterator&);
using shrink_fn = void (*)(TensorIterator&, Scalar);
using shrink_backward_fn = void (*)(TensorIterator&, Scalar);
using elu_fn = void (*)(TensorIterator&, Scalar, Scalar, Scalar);
using leaky_relu_fn = void (*)(TensorIterator&, Scalar);
using leaky_relu_backward_fn = void (*)(TensorIterator&, Scalar);
using log_sigmoid_cpu_fn = void (*)(Tensor& , Tensor&, const Tensor& );
```

**Comment by albanD** [comment]
I don't think there is a rule. Where it makes sense.
Since you're going to make something very close to sigmoid, I would recommend close to sigmoid.

Note that because the main benefit we expect here is from the autograd and not the low level implementation, I would recommend as a first step that the changes in the aten folder are limited to a new function in `native_functions.yaml` and a generic function in "Activation.cpp" that looks like (no need for cpu/gpu specific implementations):
```
Tensor swish(const Tensor & self) {
  return self * at::sigmoid(self);
}
```

The benefit from memory that you want is going to come from the fact that you're gonna register this function in `tools/autograd/derivatives.yaml` with a custom backward function!

**Comment by glaringlee** [comment]
Assign to me for now. This seems to me a good onboarding task, will assign this to onboarding people later.

**Comment by glaringlee** [comment]
Hi, @rwightman, we have a newly joined people in the team, I think this could be a good onboarding task for him, do you mind letting him doing this?

**Comment by rwightman** [comment]
@glaringlee sure no problem, it does make a good intro task

**Comment by glaringlee** [comment]
@rwightman Thank you!

**Comment by heitorschueroff** [comment]
Implemented in this PR https://github.com/pytorch/pytorch/pull/41034.

**Comment by rwightman** [comment]
@albanD @heitorschueroff FYI, I tested the SiLU on nightly with some EfficientNet networks. Everything is working fine, memory appears on par with my autograd impl. However, the throughput is approx 20% lower vs my jit + autograd version (with fusion enabled, either comared to 1.4 with the old jit or the 1.7 nightly with legacy enabled). That's probably worth looking into going the next step and impl down to CUDA level?
